### PR TITLE
Wait for Ajax requests and open dropdown menus

### PR DESCRIPTION
**Pull request by ykitamura-mdsol**
Sometimes we need to wait for Ajax requests to complete.
https://robots.thoughtbot.com/automatically-wait-for-ajax-with-capybara

Also, drop-down menus are not visible.
able to run script such as `$('.dropdown').addClass('open');` to open them to make visible and crawlable.

@mdsol/team-10

### CODE REVIEW

File: lib/grell/rawpage.rb [hunk 1]
```diff
@@ -36,6 +35,17 @@ def has_selector?(selector)
       page.has_selector?(selector)
     end
 
+    def finished_all_ajax_requests?(timeout, interval)
+      Timeout::timeout(timeout) do
```
Comment (jcarres-mdsol) [inline lib/grell/rawpage.rb:39]: This seems to be similar to what `lib/grell/reader.rb` does. Couldn't we only evaluate jquery without bothering with timing here?

File: lib/grell/crawler.rb [hunk 2]
```diff
@@ -15,6 +15,7 @@ def initialize(options = {})
       end
 
       @driver = CapybaraDriver.setup(options)
+      @run_before = options[:run_before]
```
Comment (jcarres-mdsol) [inline lib/grell/crawler.rb]: probably should be called `evaluate_in_each_page` or similar

File: lib/grell/crawler.rb [hunk 3]
```diff
@@ -56,13 +57,15 @@ def start_crawling(url, options = {}, &block)
...
       Grell.logger.info "Visiting #{site.url}, visited_links: #{@collection.visited_pages.size}, discovered #{@collection.discovered_pages.size}"
       site.navigate
+      site.rawpage.page.evaluate_script(@run_before) if @run_before
       filter!(site.links)
       add_redirect_url(site)
 
       if block # The user of this block can send us a :retry to retry accessing the page
         while crawl_block(block, site) == :retry
           Grell.logger.info "Retrying our visit to #{site.url}"
           site.navigate
```
Comment (jcarres-mdsol) [inline lib/grell/crawler.rb]: I think it is time to DRY this so we do not copy/paste the lines above

File: lib/grell/crawler.rb [hunk 4]
```diff
@@ -56,13 +57,15 @@ def start_crawling(url, options = {}, &block)
     def crawl(site, block)
       Grell.logger.info "Visiting #{site.url}, visited_links: #{@collection.visited_pages.size}, discovered #{@collection.discovered_pages.size}"
       site.navigate
+      site.rawpage.page.evaluate_script(@run_before) if @run_before
```
Comment (jcarres-mdsol) [inline lib/grell/crawler.rb]: I'm scared we will create a lot of javascript errors in the apps if they fail to evaluate the script.

**Comment by jcarres-mdsol** [comment]
Needs explanation in the readme on how this new thing works

File: lib/grell/rawpage.rb [hunk 5]
```diff
@@ -36,6 +35,17 @@ def has_selector?(selector)
...
+        (timeout / interval).ceil.times do
+          jquery_active = page.evaluate_script('jQuery.active')
```
Comment (jfeltesse-mdsol) [inline lib/grell/rawpage.rb]: if jQuery is not being used on the page the JS engine will throw

```
Uncaught ReferenceError: jQuery is not defined
```

**Comment by ssteeg-mdsol** [comment]
Seems like Travis is having lots of problems recently with getting PhantomJS - https://travis-ci.org/mdsol/grell/jobs/156411765

```
The command 
"wget https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-linux-x86_64.tar.bz2 -O $PWD/travis-phantomjs/phantomjs-2.1.1-linux-x86_64.tar.bz2" 
failed and exited with 8 during .
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,7 @@
+# 1.6.11
+  * Wait for Ajax requests to complete to load all links
```
Comment (ssteeg-mdsol) [inline CHANGELOG.md]: Maybe `Ensure all links are loaded by waiting for Ajax requests to complete` ?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,7 @@
...
+  * Added '@run_before' option to evaluate before extracting links (e.g. $('.dropdown').addClass('open');)
```
Comment (ssteeg-mdsol) [inline CHANGELOG.md]: Add*
Should keep everything in the same tense.

File: lib/grell/rawpage.rb (hunk 1 above)
Comment (ykitamura-mdsol) [reply lib/grell/rawpage.rb:39]: It doesn't work without `Timeout` here.
For some reason `page.evaluate_script('jQuery.active')` never returns 0 if `Timeout` is removed

File: lib/grell/crawler.rb (hunk 4 above)
Comment (ykitamura-mdsol) [reply lib/grell/crawler.rb]: I think the errors will occur inside Capybara. Does it affect NewRelic or other monitoring systems?

File: lib/grell/rawpage.rb (hunk 5 above)
Comment (ykitamura-mdsol) [reply lib/grell/rawpage.rb]: `page.evaluate_script` just returns `nil` when an exception raises in the JS engine

File: README.md [hunk 8]
```diff
@@ -151,6 +151,17 @@ The page object generated by accessing the first URL passed to the start_crawlin
...
+
+You can evalute a JavaScript snippet in each page before extracting links by passing the snippet to the 'evaluate_in_each_page' option:
+
+```ruby
+require 'grell'
+
+crawler = Grell::Crawler.new(evaluate_in_each_page: "$('.dropdown').addClass('open');")
+
+```
+
```
Comment (ykitamura-mdsol) [inline README.md:164]: added some explanation

File: lib/grell/rawpage.rb (hunk 5 above)
Comment (jcarres-mdsol) [reply lib/grell/rawpage.rb]: but we will be creating javascript errors on the page, it depends on who would  not like that

File: lib/grell/rawpage.rb (hunk 1 above)
Comment (jcarres-mdsol) [reply lib/grell/rawpage.rb:39]: reader runs everything inside a timeout itself, so we have a timeout inside a timeout...

File: lib/grell/page.rb [hunk 9]
```diff
@@ -26,11 +26,12 @@ def initialize( url, id, parent_id)
     end
 
     def navigate
-      # We wait a maximum of WAIT_TIME seconds to get an HTML page. We try or best to workaround inconsistencies on poltergeist
+      # We wait a maximum of WAIT_TIME seconds to get an HTML page. We try our best to workaround inconsistencies on poltergeist
       Reader.wait_for(->{@rawpage.navigate(url)}, WAIT_TIME, WAIT_INTERVAL ) do
         @rawpage.status && !@rawpage.headers.empty? &&
           @rawpage.headers["Content-Type"] && @rawpage.headers["Content-Type"].include?('text/html').equal?(true)
       end
+      @rawpage.finished_all_ajax_requests?(WAIT_TIME, WAIT_INTERVAL)
```
Comment (jcarres-mdsol) [inline lib/grell/page.rb]: Maybe better call this `wait_for_all_ajax_requests`

**Comment by ykitamura-mdsol** [comment]
addressed comments and added specs

**Comment by jfeltesse-mdsol** [comment]
@ssteeg-mdsol maybe we should host that file ourselves, @jcarres-mdsol is there such a thing as a file sharing folder or something apps can access during deployment?

File: lib/grell/rawpage.rb [hunk 10]
```diff
@@ -36,6 +35,17 @@ def has_selector?(selector)
       page.has_selector?(selector)
     end
 
+    def wait_for_all_ajax_requests(timeout, interval)
+      Timeout::timeout(timeout) do
+        (timeout / interval).ceil.times do
+          jquery_active = page.evaluate_script('(typeof(jQuery)!="undefined") && jQuery.active;')
```
Comment (jfeltesse-mdsol) [inline lib/grell/rawpage.rb]: this should be `!==` and btw typeof doesn't need `()` so you could write

```
typeof jQuery !== "undefined" && jQuery.active;
```

(untested in this context but it's legit JS syntax)

Comment (ykitamura-mdsol) [reply lib/grell/rawpage.rb]: Thanks! modified it

**Comment by thoshikawa-mdsol** [comment]
I found the [thread](https://github.com/travis-ci/travis-ci/issues/3225) which is talking about PhantomJs in Travis.

According to that, `Bitbucket` is blocking download. 
For the temporary solution while waiting for hosting ourselves, we can use following download source.

https://assets.membergetmember.co/software/phantomjs-2.1.1-linux-x86_64.tar.bz2

**Comment by jcarres-mdsol** [comment]
@jfeltesse-mdsol @thoshikawa-mdsol I do not think we have a good way to solve this. We could make a github repo hosting that file. Phantomjs has around 1 release per year, should be low maintenance

**Comment by jcarres-mdsol** [comment]
Voilá, we can use this instead: https://github.com/JordiPolo/phantomjs/blob/master/phantomjs-2.1.1-linux-x86_64.tar.bz2?raw=true

**Comment by ykitamura-mdsol** [comment]
It works. Thank you Mr. Polo

**Pull request merged**